### PR TITLE
Move from Grunt to Gulp+libsass (really fast).

### PR DESCRIPTION
**Pull request by bserem**
```
Move source files under source/
Produce dist files under dist/
Symlink bootstrap JS files so that get parsed by JS minification (better than Drupal minification).
Create placeholder folder for several assets, they are designed to work with Gulp and offer
all kinds of magic (retina logos, imagesprites, image minification, google webfonts local copies etc).
Custom webfont creation from SVG files is still a work in progress.
```

The Gulpfile needs to be cleaned up further in order for people to easily understand it. I'm planning to do so soon.
Right now, simply using "gulp watch" and "gulp production" will allow you to work while developing and once finished will minify everything and clean+combine what it can to save as much as possible.

I'll be working with this changes in the upcoming days, so more commits are to follow, plus improved instructions.

Sadly, I do not have enough time for custom webfonts, which are the killer feature of this approach.
It can be done, I've done it, but it requires some time to fully automate it.

### CODE REVIEW

**Comment by bserem** [comment]
Important:
this https://www.drupal.org/node/2329453
is required for drupal to work when more .info files exist (because of node+bower).

I've been using patch #8 for many many months, until this gets resolved. It seems that the maintainers are closer to a better fix.

I believe this is irrelevant of the PR, since the target branch also ads .info files for bower.

**Comment by wrender** [comment]
Hi bserem. Thanks so much for looking into this.  My initial idea was to make this compatible with as many compilers as possible.  So the idea is to make it work with Compass, Grunt and Gulp...  This way people on different platforms can use it.  For example if you are on an older system sometimes access to Ruby/Compass easier.  Do you have any thoughts on this?

Also, would not not make more sense to have Drupal compile the JS.  This way third party modules like CSS/JS Advanced Aggregation would work properly?

**Comment by bserem** [comment]
I have no experience with Compass (at all), I went with libsass right away because it is way faster ad it made more sense to me back then. But I suppose that simply adding the Gulpfile (without removing the stuff I removed in my PR) will add Gulp+libsass compatibility.

Regarding CSS/JS, my Gulp configuration does a better job at minifying the files than Drupal. It is not my doing, the Gulp plugins I use are better in this task than Drupal.
Aggregation continues to work properly as long as the CSS/JS files are provided inside the info file (and not inside .tpl.php files).

As an extra bonus, with my approach on JS, one can remove bootstrap JS features that doesn't need (for example tooltips). If somebody needs the full bootstrap JS then he/she is better getting it from CDN anyway.

On a big project I've been working in the past months I do the same with the bootstrap CSS too: I remove the CSS I do not require.
For example, I completely remove panels, progress-bars, glyphicons, popovers and a bunch of other bootstrap features, thus making the final CSS file (including my changes on top) only a fragment of the original filesize.
I have plans to add such functionality on this PR too, but I need to understand where you are going with this, so that I don't break something. My PR is already a bit "too aggresive" :)
This is mainly because it is a product of many weeks of development, which I've promised to a lot of friends to release before the year ends!

Want me to do some changes towards the PR, apart from allowing it to live side by side with the other two platforms?
I strongly suggest to use the "source/dist" directories, since it makes things so much easier when it comes to version control and it also allows one to do a fully clean start by deleting the "dist" folder.

**Comment by wrender** [comment]
Thanks for the detailed explanation.  I'm just going to setup a vagrant machine this morning so I can test this out and make sure it works with Compass, and Grunt for compiling.  I will also review your updates to the documentation.  I will then commit it.

I agree; being able to easily comment out unused Bootstrap components is a great idea.

Yes. If you can do those changes it would be appreciated.  I think we should work towards putting a release on Drupal.org?  My idea of this starter theme is basically to take a different direction from the Drupal Bootstrap Project, in that this startertheme can be easily customized like the Zen or Omega project.  After talking to the maintainers of the Drupal Bootstrap project he mentioned they are only going to keep that project developing towards beginner website users. (Not theme developers).  The only features I find that are useful in their project is the Drupal theme hooks to alter core CSS to match Drupal.. (Which they've done a great job on).

**Comment by bserem** [comment]
My PR currently breaks Grunt/Compass. We're (I'm) gonna have to restore them. Not sure if there is enough time to prepare this withing the year though!

I suppose that simply having the official drupal bootstrap project as a base is enough for most people.
Cloning this project will allow them to do all sorts of "magic stuff" more easily and will certainly provide more power to theme developers (who have no need for CDN).
This was my approach when I starting working with bootstrap and Drupal, and it has been working fine for me till now.

**Comment by wrender** [comment]
Ok. Right, I forgot that it is new years eve today!

Yes. That is the approach I was using, is just quickly put together my own Sass starterkit with the Drupal Bootstrap project for each website we were building, but I found that I wasted so much time doing that.

Do you push commits to Drupal.org?  The reason I ask is I have also started a sandbox here: https://www.drupal.org/sandbox/wrender/2627564  .  I think it would be good for us to put all our efforts on Drupal.org for the benefit of the Drupal community.

**Comment by bserem** [comment]
New years eve isn't an issue usually, but I am the host this time :)

Yeah, I have project rights in drupal.org, it doesn't allow for forking/PRs though, only patches.

**Comment by wrender** [comment]
Ok. I've added you to the Drupal.org sandbox project.  Travisc asked me to move this to github for him for some reason, but I think it would benefit more if we keep it up to date on drupal.org too...

**Comment by wrender** [comment]
Great work on this. I've got a bit of feedback:

1.  I think we should closely follow the directory structure from the Drupal documentation: https://www.drupal.org/node/2349803.   And also use the SMACSS style organization directly from here: https://www.drupal.org/node/1887922  (I will start working on re-organizing this today/tomorrow.  If you look at the Zen subtheme it follows this structure and is quite easy to use)

2.  In your pull request you have included the js files. Since this are part of the Bootstrap-Sass package I think they should really not be included with the project.

3.  I don't think we should remove Grunt. Gulp is really just an alternative to Grunt for compiling. Ideally I think having more compiler support really helps for developers working in all kinds of environments/servers.  Can you add gruntfile.js back for this reason?  (I just noticed that the compass config.rb folder needs fixing too. I will do this.)

4.  Since Gulp is a different compiler we should separate out the instructions in the documentation for this from the grunt section. (To avoid people being confused and think they need to use two different compilers)

**Comment by wrender** [comment]
I made some commits to the sanbox project on Drupal.org.  I've moved the sass folder into a SMACCS recommended configuration.

Still not sure about the compilers. It seems like all of the compilers: compass, grunt, gulp, sass all conflict with one another.  So I'm thinking it might be best to just stick with the most widely used NPM one: grunt.  Any thoughts on this?

**Comment by bserem** [comment]
1. Agreed. Foundation also has a very nice SMACSS directory structure in their project.

2. The JS files are not included, they are symlinks. Git allows symlinks.
They point to the location of the JS files of bootstrap. Removing a symlink will remove that functionallity from your project, without you losing the original JS file.

3. Gulp is way easier to comprehend/configre than Grunt. It is also a newer technology. Gulp allows the use of both libsass or compass for compiling SASS. So, you do not necessarily need Compass itself, you can call it from Gulp.

4. Agreed, if we end keeping both of them.

**Comment by wrender** [comment]
I don't see the point in having this project on both Github and Drupal.org.   Would you be able to work on it at Drupal.org?   I understand that Github is a bit nicer to work with, but I think for the benefit of the drupal community it would be good to just keep it there.  

Do you want to go ahead and add gulp build file to the project at https://www.drupal.org/sandbox/wrender/2627564  ?

If you don't mind, just leave the symlinks out for now, and keep the directory structure I have in place:
/sass
/css
/bootstrap-sass/

**Comment by wrender** [comment]
I've synced over my changes that I worked on the last couple of days from drupal.org.  So if you prefer to work on github the master here is now up to date.  I still think maybe it would be better to just work from drupal.org sandbox repo though?

**Comment by wrender** [comment]
I've synced over the recent changes that I did the last couple of days from drupal.org sandbox repo.  Just in case you wanted to work from github.  I'm still thinking it would be better for us to try and just work from the drupal.org repo, as that way maybe our project could get approved and help the drupal community more.  Cheers.

**Comment by arttus** [comment]
I think the project would be better served by using Gulp. If you look at the trend-line it's the direction people are going. 

That being said, having both just costs cognitive load. It’s better to have an opinion and use best practices for that one technique.

If you want to use Grunt fine, I just would just avoid trying to munge both together.

**Comment by wrender** [comment]
Ok. Thanks for the feedback.  Sure. Let's move the Gulp since the community seems to be going in that direction and you guys prefer it.  There is a good article here on that here: http://sixrevisions.com/web-development/grunt-vs-gulp/     From what I can see it is comparing two types of similar apples.  I think the industry seems to be moving away from Compass/Ruby so let's just drop support for that for now unless we have someone come along that really needs it on a large scale.

bserem .  You're gulp build has a lot of good stuff in it.  I think we should really strip it down, and document it like you mentioned too.  We need to ask, is this a really important task that every theme developer would use?  I would suggest we can do this on Drupal.org by creating feature requests under the project.  For example, one feature would be to "symlink to disable unused javascript bootstrap features", etc...  

travis pointed out that browsersync is a really important feature.  I agree, I just tested it out with grunt a couple days ago, and it drastically speeds up theme development by automating page reloads.

I'm going to scrap this project on Github for the sake of the Drupal community. Please read here for my reasons: https://www.drupal.org/node/648898

Then hopefully the maintainer of the Bootstrap of Drupal project would come to his senses and eventually merge in our project.

**Comment by bserem** [comment]
Regarding my Gulp config, one can call only the functions he needs, and there is a logic behind all these lines in the gulpfile!

I need to document them inside the file.
In short:
watch and production are compatible, they only differ in js/css being outputed
images/sprites/logos are automatically compressed to get an A mark in pagespeed and yslow
retina sprites are automatically created, if they exist
font files are automatically copied to 'dist' folder (remember, dist is not under version control)

So, while we could strip things a bit, one would have to decide which of them are to be removed.

Now, regarding browsersync, I've been using livereload successfully, on remote development servers, which reloads only the CSS and not the whole page. It is fast, it is helpfull and it is pain to install in Drupal :(
If you guys have experience with browsersync on remote environments I'm all ears. But I was under the assumption that browsersync only works under localhost.
Eitherway, if you point me to its documendation I can probably intergrate it in gulp watch task.

Regarding drupal.org, lets continue there as far as this project is concerned.

**Comment by wrender** [comment]
Good morning bserem.  Your comments are great. Thanks for getting back to me.

I've added a basic gulpfile.js to https://www.drupal.org/sandbox/wrender/262756 as a baseline for us to add tasks to.  (I've already updated the documentation there too and added support for Browsersync).

Would you be able to open separate issues on drupal.org for any gulpfile.js tasks you want to add?

Regarding some of the tasks: 
1.   Logo  -  Drupal already provides a theme function for uploading a logo to the theme. If a retina logo is to be added to the theme and template files, then a theme function should be created to disable the logo upload option in the theme to avoid confusion by the website administrator?

2.  CSS/JS Compression -  There is already a module that handles this.  "Advanced CSS/JS Aggregation".  I don't see the point in doing compression here.  It is just extra unnecessary tasks that make development even more confusing by more task lines to review.

3.  Google Web Fonts -  I think this is a really good idea to add.

4.  Sprites - Does this really need to be included?

5.  The task "Clean-Up Dist", "Copy files to dist as needed", "Prepare bootstrap sass subtheme" seem potentially unnecessary.

Regarding your problems using Browsersync.  I have not come across any problems yet. Although I am using dockers and vagrant with mostly local development on Ubuntu.  I will test setting up a browsersync session on a centos/cpanel development server and see if everything works ok.

We thoroughly tested the most recent code on drupal.org over the weekend, and couldn't find any issues so far.

**Comment by wrender** [comment]
I tested using browsersync from the drupal project on a cPanel/Cloudlinux/CentOS Development server.  It works properly.  The main things you need to remember when using browsersync on a remote server:
-  You need to set the proxy entry to the correct hostname/ip of your drupal website
-  You need to be able to access port 3000 on the server. (In my cPanel firewall I had to open this up), and then re-run "gulp browsersync".
-  You need to navigate to the "External" address with port 3000 on the end.

I'm not sure if you are familiar with vagrant, or virtual machines.  But someone has developed a really good local drupal development system here:  http://www.drupalvm.com/

I would imagine that most people are going to be moving to dockers though, as it is way more efficient virtualization/local drupal development friendly.

**Comment by bserem** [comment]
I am not familiar with vagrant, nor browsersync. I'll keep an eye on them though.

Regarding gulp. Remember that my approach is for an advanced subtheme, and a way too advanced gulpfile:

1) Logo: I do not use the drupal logo. I add all logo variations into sprites, have them compressed by superior image minification tools, and send a sprite to the client. My projects often use two variations of the logo (mobile/desktop) and if you add retina to it you go to four. I can solve this by using one CSS class and media queries (easy) instead of fighting with drupal.

2) Minifying CSS/JS on the theme level doesn't bypass ADVAGG. Plus, I find that I have better output through gulp (combined media queries, combined selectors etc) which ADVAGG can't handle yet. Advagg is great, and we use it a lot. It can handle modules css/js aggregation and it can inline CSS, but it can't offer the compression I get by css-nano for example.

3) Google web fonts is there because it allows you to get a local copy of the webfont, and get a higher mark on pagespeed. It is cool, especially if you don't use apache to add mod-pagespeed (which will also mess with your static caches)

4) I find this to be the most important aspect. Managing sprites is a pain. What I do here is that all files that are in a specific folder get minified and combined into a sprite. A CSS class is assigned to each one, that remains the same as long as the filename doesn't change. So, if you add a sprite in the future, your theme will not be affected. I also use it with logos, since my logos get sent in the form of sprites (example: http://www.solebich.de/sites/www.solebich.de/themes/s2/dist/assets/images/sprites.png). After all, all of this has been worked and is being done automatically. All you need to do is add the proper css class wherever you want.

5) These tasks need a bit of polishing. A clean-up is good to have, and it is good to be piped before every other task. Preparation tasks can be used in combination with firstrun. You do not see the full power of it here, but I've also added drush commands in my project, that enable/disable aggregation based on whether you are working on dev or on production. I could also add bower in there, and automate this too!
In short, they need to be adapted.
